### PR TITLE
Fixes #12235: handle dangling reference during VHDS update request

### LINKED ISSUE #12235: Potential Dangling Reference in On Demand VHDS
*Title*: *Potential Dangling Reference in On Demand VHDS*

*Description*:
> In the call stack of [on-demand vhds](https://github.com/envoyproxy/envoy/blob/bc02b7748a69fd3854f74b9e5bec236b5499f7fd/source/common/router/rds_impl.cc#L310),  this pointer is captured in the callback lambda.
subscription_->updateOnDemand(alias);
In the statement above,  subscription_ is actually a reference to a shared pointer, whose lifetime is dynamic. If it has been destroyed before the callback is run in main thread, this will become a dangling reference and the program may crash. I suggest capturing a weak_ptr in the closure for safety.

### PR DESCRIPTION
**Pull request by dmitri-d**
Fixes https://github.com/envoyproxy/envoy/issues/12235

Signed-off-by: Dmitri Dolguikh <ddolguik@redhat.com>

Risk Level: low
Testing: existing tests
Docs Changes: none
Release Notes: none

### CODE REVIEW

**Comment by dmitri-d** [comment]
ping @stevenzzzz

File: source/common/router/rds_impl.cc [hunk 1]
```diff
@@ -306,11 +306,19 @@ void RdsRouteConfigProviderImpl::requestVirtualHostsUpdate(
...
       VhdsSubscription::domainNameToAlias(config_update_info_->routeConfigName(), for_domain);
-  factory_context_.dispatcher().post([this, alias, &thread_local_dispatcher,
-                                      route_config_updated_cb]() -> void {
-    subscription_->updateOnDemand(alias);
-    config_update_callbacks_.push_back({alias, thread_local_dispatcher, route_config_updated_cb});
-  });
+  // We use weak pointers here as RdsRouteConfigProviderImpl instance could go away before the
+  // dispatcher had a chance to execute the callback.
+  factory_context_.dispatcher().post(
+      [sub = std::weak_ptr<RdsRouteConfigSubscription>(subscription_),
```
Comment (stevenzzzz) [inline source/common/router/rds_impl.cc:312]: I think you only need the config_cbs to help decide if the config provider is gone.

File: source/common/router/rds_impl.cc [hunk 2]
```diff
@@ -306,11 +306,19 @@ void RdsRouteConfigProviderImpl::requestVirtualHostsUpdate(
...
+       config_cbs = std::weak_ptr<std::list<UpdateOnDemandCallback>>(config_update_callbacks_),
+       alias, &thread_local_dispatcher, route_config_updated_cb]() -> void {
+        auto s = sub.lock();
+        auto cb = config_cbs.lock();
```
Comment (stevenzzzz) [inline source/common/router/rds_impl.cc:316]: use a better name pls?

File: source/common/router/rds_impl.cc [hunk 3]
```diff
@@ -306,11 +306,19 @@ void RdsRouteConfigProviderImpl::requestVirtualHostsUpdate(
...
+        if (s && cb) {
+          s->updateOnDemand(alias);
+          cb->push_back({alias, thread_local_dispatcher, route_config_updated_cb});
```
Comment (stevenzzzz) [inline source/common/router/rds_impl.cc:319]: if in main thread, management server already sent back response with the alias, do you want to send another on demand request to management server?

Depends on management server implementation, if it only sends a response when there is some update of the project, then this HttpRequest will hang till timeout, while Envoy has the config to proceed.

File: source/common/router/rds_impl.cc (hunk 2 above)
Comment (dmitri-d) [reply source/common/router/rds_impl.cc]: renamed

File: source/common/router/rds_impl.cc (hunk 1 above)
Comment (dmitri-d) [reply source/common/router/rds_impl.cc]: Yes, but it's dependent on the position of the member var in the source file. I figured this is the safest?

File: source/common/router/rds_impl.cc (hunk 3 above)
Comment (dmitri-d) [reply source/common/router/rds_impl.cc]: This gets pretty complicated quickly: the only time (that I can see) an instance of `RdsRouteConfigProviderImpl`can be deallocated is when LDS is used and the listener configuration has changed. At this point all connections are going to be drained, trying to update a vhost seems not worth it to me.

Even if we wanted to do that, we'd still need an active subscription (may not be there yet, and no way to find out which one to use) and a place to store the callback to notify the on_demand_update filter (which may not be there any more, as the active stream it belongs to may have been closed).

**Comment by dmitri-d** [comment]
* added a test

**Comment by dmitri-d** [comment]
* merged in changes from master

File: source/common/router/rds_impl.cc (hunk 1 above)
Comment (stevenzzzz) [reply source/common/router/rds_impl.cc]: I dont get it: as you pointed out in the other comment, the RDS provider is guaranteed to be allocate/deallocated on main thread. so if we can lock config_cbs here, that means the subscription is definitely alive. this lambda is run in main thread, where deallocation of RDS provider happens in sequence with this lambda.

File: source/common/router/rds_impl.cc (hunk 3 above)
Comment (stevenzzzz) [reply source/common/router/rds_impl.cc]: I am describing a scenario where no rds subscription destruction is involved. 

could the management server send required vhost to Envoy before the on-demand request is sent? in that case, the management server may or may not send the requested resource again depends on its implementation, in which case the callback will not be called and the request will only timeout.  this is very implementation specific, so I am not too concerned.

File: test/common/router/rds_impl_test.cc [hunk 4]
```diff
@@ -278,6 +279,39 @@ TEST_F(RdsImplTest, FailureSubscription) {
...
+  testing::NiceMock<Event::MockDispatcher> local_thread_dispatcher;
+  {
+    auto rds = RouteConfigProviderUtil::create(
+        parseHttpConnectionManagerFromYaml(rds_config), server_factory_context_,
+        validation_visitor_, outer_init_manager_, "foo.", *route_config_provider_manager_);
+
+    EXPECT_CALL(server_factory_context_.dispatcher_, post(_)).WillOnce(SaveArg<0>(&post_cb));
+    rds->requestVirtualHostsUpdate("testing", local_thread_dispatcher,
+                                   std::make_shared<Http::RouteConfigUpdatedCallback>(
+                                       Http::RouteConfigUpdatedCallback([](bool) -> void {})));
```
Comment (stevenzzzz) [inline test/common/router/rds_impl_test.cc:306]: nit: might worth check if the lambda is actually not called. 

The dangling reference may not lead to an Exception but some segv error.

File: source/common/router/rds_impl.cc (hunk 1 above)
Comment (dmitri-d) [reply source/common/router/rds_impl.cc]: You are right, fixed

**Comment by dmitri-d** [comment]
ping @stevenzzzz.

File: source/common/router/rds_impl.cc (hunk 1 above)
Comment (stevenzzzz) [reply source/common/router/rds_impl.cc]: sorry, I think you want to check the config_cbs_ actually. 
Subscription is shared between routeconfigProviders, there could be a case where this route provider could be gone, but the subscription stays.

**Comment by stevenzzzz** [comment]
sorry I thought I committed the comment.

there is one thing to be fixed around which weak_ptr to look at, please lock the most "specific" object to make sure there is no lifecycle problem, i.e., the config_cbs_.

**Comment by dmitri-d** [comment]
> there is one thing to be fixed around which weak_ptr to look at, please lock the most "specific" object to make sure there is no lifecycle problem, i.e., the config_cbs_

In this context, a weak_ptr to either the subscription or the list of callbacks would work; if anything the subscription is longer-lived and has a benefit of being already stored in a shared_ptr.

**Comment by dmitri-d** [comment]
* switched to use config_update_callbacks to determine if the instance of RdsRouteConfigProviderImpl is still alive.

File: source/common/router/rds_impl.cc [hunk 5]
```diff
@@ -306,11 +306,19 @@ void RdsRouteConfigProviderImpl::requestVirtualHostsUpdate(
...
-    subscription_->updateOnDemand(alias);
-    config_update_callbacks_.push_back({alias, thread_local_dispatcher, route_config_updated_cb});
-  });
+  // We use weak pointers here as RdsRouteConfigProviderImpl instance could go away before the
+  // dispatcher had a chance to execute the callback.
+  factory_context_.dispatcher().post(
+      [subscription = std::weak_ptr<RdsRouteConfigSubscription>(subscription_),
+       config_cbs = std::weak_ptr<std::list<UpdateOnDemandCallback>>(config_update_callbacks_),
+       alias, &thread_local_dispatcher, route_config_updated_cb]() -> void {
+        auto sub = subscription.lock();
```
Comment (stevenzzzz) [inline source/common/router/rds_impl.cc:315]: you can move this into the if statement below.

**Comment by stevenzzzz** [comment]
> > there is one thing to be fixed around which weak_ptr to look at, please lock the most "specific" object to make sure there is no lifecycle problem, i.e., the config_cbs_
> 
> In this context, a weak_ptr to either the subscription or the list of callbacks would work; if anything the subscription is longer-lived and has a benefit of being already stored in a shared_ptr.
 if config_callbacks are gone, it's undefined behavior again when you push a callback into the list.

Thanks for the fix.
LGTM

**Comment by dmitri-d** [comment]
* responded to feedback

File: source/common/router/rds_impl.cc (hunk 5 above)
Comment (stevenzzzz) [reply source/common/router/rds_impl.cc:315]: could you also move this sub lock to the if statement below?

**Comment by dmitri-d** [comment]
* responded to feedback

**Comment by dmitri-d** [comment]
Looks like 1rm build failure is due to a flaky test.

**Comment by dmitri-d** [comment]
ping @stevenzzzz

**Comment by stevenzzzz** [comment]
I LGTMed already.
Let's ping @snowp   
😄

**Comment by dmitri-d** [comment]
* responded to feedback
* merged in master

ping @snowp

**Comment by stevenzzzz** [comment]
/assign htuch@ 
FYI Harvey

**Comment by stevenzzzz** [comment]
/assign @htuch 
FYI

File: source/common/router/rds_impl.cc [hunk 6]
```diff
@@ -306,11 +306,17 @@ void RdsRouteConfigProviderImpl::requestVirtualHostsUpdate(
     std::weak_ptr<Http::RouteConfigUpdatedCallback> route_config_updated_cb) {
   auto alias =
       VhdsSubscription::domainNameToAlias(config_update_info_->routeConfigName(), for_domain);
-  factory_context_.dispatcher().post([this, alias, &thread_local_dispatcher,
-                                      route_config_updated_cb]() -> void {
-    subscription_->updateOnDemand(alias);
-    config_update_callbacks_.push_back({alias, thread_local_dispatcher, route_config_updated_cb});
-  });
+  // The RdsRouteConfigProviderImpl instance can go away before the dispatcher has a chance to
+  // execute the callback. We use a weak_ptr here to verify that the instance is still around.
```
Comment (mattklein123) [inline source/common/router/rds_impl.cc:310]: Why don't we allocate the provider itself with shared_from_this and then we can capture a weak reference to ourselves and post that? This seems a bit strange and fragile?

Comment (stevenzzzz) [reply source/common/router/rds_impl.cc:310]: I'd hesitate to do so, with that we could get into another rabbit hole where a worker thread may accidentally get a shared_ptr of the provider and introduce another bunch of life cycle problem due to provider can only be con/destructed on main thread. 
thoughts?

Comment (mattklein123) [reply source/common/router/rds_impl.cc:310]: OK fair enough. Can we add a comment on why we are not doing that so the next person that comes along doesn't ask the same question? Thank you.

/wait

Comment (dmitri-d) [reply source/common/router/rds_impl.cc:310]: Ahh, yes, this is better.

**Comment by dmitri-d** [comment]
* switched to using a shared_from_this instance of a RdsRouteConfigProviderImpl
 * merged in latest master

File: source/common/router/rds_impl.h [hunk 7]
```diff
@@ -193,7 +193,8 @@ struct UpdateOnDemandCallback {
  * the subscription.
  */
 class RdsRouteConfigProviderImpl : public RouteConfigProvider,
-                                   Logger::Loggable<Logger::Id::router> {
+                                   Logger::Loggable<Logger::Id::router>,
+                                   public std::enable_shared_from_this<RdsRouteConfigProviderImpl> {
```
Comment (stevenzzzz) [inline source/common/router/rds_impl.h:197]: can we please not do this. When provider is capable of shared_from_this, we actual open its access to worker threads, which I'd hope we can avoid: providers are supposed to be created/owned by HCM and no one else should have ownership to it, otherwise it'd introduce lots of trouble to do destruction orders when Envoy tears down. 
If you need a cleaner signal to decide if we need a post, one thing you can do is to lock on route_config_updated_cb, which the finest grained object that tells if the source httprequest is alive or not.

File: source/common/router/rds_impl.cc [hunk 8]
```diff
@@ -308,15 +308,16 @@ void RdsRouteConfigProviderImpl::requestVirtualHostsUpdate(
...
-       config_cbs = std::weak_ptr<std::list<UpdateOnDemandCallback>>(config_update_callbacks_),
-       alias, &thread_local_dispatcher, route_config_updated_cb]() -> void {
-        if (auto callbacks = config_cbs.lock()) {
-          subscription_->updateOnDemand(alias);
-          callbacks->push_back({alias, thread_local_dispatcher, route_config_updated_cb});
-        }
-      });
+  factory_context_.dispatcher().post([this,
+                                      self = std::weak_ptr<RdsRouteConfigProviderImpl>(
+                                          shared_from_this()),
```
Comment (stevenzzzz) [inline source/common/router/rds_impl.cc:313]: this could cause trouble as it introduces another race between worker threads and main thread: for example on envoy teardown, the lambda created in the worker thread may be the last one to be teardown, in which case, the destruction of the provider will happen in worker thread, which triggers an assert failure because it holds an TLS instance which can only be destructed in main thread.

Comment (mattklein123) [reply source/common/router/rds_impl.cc:313]: But how is this safe then? We capture "this" so it can still crash?

File: source/common/router/rds_impl.h (hunk 7 above)
Comment (mattklein123) [reply source/common/router/rds_impl.h:197]: I don't quite understand this now that I think about it more. This is already a shared_ptr allocated object. Why does it matter if it enabled shared_from_this?

File: source/common/router/rds_impl.cc (hunk 8 above)
Comment (stevenzzzz) [reply source/common/router/rds_impl.cc:313]: I think we want it shared between listeners in main thread, but its ownership should not be shared to worker thread.
If however the ownership of the provider is shared to worker thread, the scenario i was talking about would cause the assertion error when LDS deletes the listener or when envoy quits.

File: source/common/router/rds_impl.h (hunk 7 above)
Comment (stevenzzzz) [reply source/common/router/rds_impl.h:197]: I think we want it shared between listeners in main thread, but its ownership should not be shared to worker thread.
If however the ownership of the provider is shared to worker thread, the scenario i was talking about would cause the assertion error when LDS deletes the listener or when envoy quits.

Comment (dmitri-d) [reply source/common/router/rds_impl.h:197]: I think the change is quite safe: we aren't changing provider interface, just the concrete implementation: to create another shared_ptr one would need to cast a RouteConfigProvider object to RdsRouteConfigProviderImpl before being able to do so. Code review would catch such a cast or a call to `shared_from_this()`. I can add a note re: intended usage of shared_from_this if that would help.

File: source/common/router/rds_impl.cc (hunk 8 above)
Comment (mattklein123) [reply source/common/router/rds_impl.cc:313]: > i am thinking capturing a member of the provider will guarantee that the provider is alive when this cb is called inside main thread.

It definitely does not guarantee that. It just prevents the member from going away.

File: source/common/router/rds_impl.h (hunk 7 above)
Comment (mattklein123) [reply source/common/router/rds_impl.h:197]: I agree that this is the right change modulo any other lifetime issue.

Comment (stevenzzzz) [reply source/common/router/rds_impl.h:197]: it's a tricky case and it's not about life cycle of provider when the post lambda gets called.
 the problem is that The ownership of any "tls instance owner" should not be spilled into other threads than the main thread.
please see my other reply why I am against this, your change is introducing another race which could ends in an assertion failure.

File: source/common/router/rds_impl.cc (hunk 8 above)
Comment (stevenzzzz) [reply source/common/router/rds_impl.cc:313]: let me rephrase why this change is a regression: 
  1. this lambda "L" is created with a shared pointer to the RDS config provider, and it's in worker thread.
  2. it's destructed in worker thread after it's posted to main thread, 
  3. in main thread, the rds config provider could go away because of LDS or Envoy teardown, in which case the ownership of the config provider will fall to the "L" if it's not deleted from the worker thread yet.
  4. deallocation of the RDS config provider in worker thread will cause an assertion failure, which is our fence on clear boundary between config pipeline and data plane, as the tls instance held by the provider can only be destructed in main thread.

Comment (mattklein123) [reply source/common/router/rds_impl.cc:313]: Yes I agree ^ is a problem, but this code is still broken. If the provider goes away on the main thread, we still capture this. That is not correct. I think given how long we are talking about this we probably need a cleaner solution here. If the original change is actually safe can the comments be clearly updated to why?

Comment (stevenzzzz) [reply source/common/router/rds_impl.cc:313]: Right. So if we capture a weak_ptr of the callback list, and verify it's alive in main thread, we can make sure that we are accessing an active provider(and its subscription), meanwhile, the provider ownership is still on main thread and we dont need to worry about the race mentioned above.

trying to lock on the route_config_updated_cb will also do the trick, but it has a different semantic: if the callback is alive, which means the hcm is alive and the provider associated with it will be alive. 
but now we are firing an on-demand when the requesting HttpRequest is alive, if we lock on provider, we are firing as long as there was a request(may be gone).

Comment (mattklein123) [reply source/common/router/rds_impl.cc:313]: Yes, OK, I think this make sense. @dmitri-d can we revert back to what you had before but also add a lot more comments? I would like to get this merged as I'm dealing with additional issues in this area and I want to fix on top of this depending on if anything else needs fixing. Thank you!

/wait

**Comment by mattklein123** [comment]
I just checked and the previous version of this patch doe *not* fix https://github.com/envoyproxy/envoy/issues/12235#issuecomment-671634502, so I will continue to debug.

**Comment by mattklein123** [comment]
Nevermind regarding ^ this is the same TSAN issue that is fixed with instrumented TSAN which I am not using. :(

File: source/common/router/rds_impl.cc (hunk 8 above)
Comment (htuch) [reply source/common/router/rds_impl.cc:313]: Do we need to capture a weak_ptr to both subscription and callback list? Or would just subscription work? I think the original solution looks right, but it seems awkward to have to lock two objects. Better documentation on lifetime guarantees and relationships would definitely help. I think what you really want is some sentinel member element that provides "am I alive?"; normally this could be `shared_from_this` but this doesn't work for the reasons @stevenzzzz points out.

**Comment by dmitri-d** [comment]
*  Reverted shared_from_this changes
* Introduced a dedicated 'still alive' flag: functionally the same as relying on a weak_ptr to config_update_callbacks_, but reads a bit clearer (I think). If everyone is ok with this, I'll add a comment explaining the lifecycle there.

**Comment by stevenzzzz** [comment]
> This makes sense to me, @stevenzzzz @mattklein123 ?

LGTM, making a shared_ptr on a callback list indeed looks weird.

**Comment by dmitri-d** [comment]
* fixed formatting
* added a comment explaining the need for `still_alive` flag
